core/server/web/api/app.js

```diff
@@ -14,16 +15,27 @@ module.exports = function setupApiApp() {
     }
 
     // If there is a version in the URL, and this is a valid API URL containing admin/content
-    // Then 307 redirect (preserves the HTTP method) to a versionless URL with `accept-version` set.
-    apiApp.all('/:version(v2|v3|v4|canary)/:api(admin|content)/*', (req, res) => {
-        const {version} = req.params;
-        const versionlessURL = req.originalUrl.replace(`${version}/`, '');
-        if (version.startsWith('v')) {
-            res.header('accept-version', `${version}.0`);
-        } else {
-            res.header('accept-version', version);
+    // Rewrite the URL and add the accept-version & deprecated headers
+    apiApp.all('/:version(v2|v3|v4|canary)/:api(admin|content)/*', (req, res, next) => {
+        let {version} = req.params;
+        const versionlessUrl = req.url.replace(`${version}/`, '');
+
+        // Always send the explicit, numeric version in headers
+        if (version === 'canary') {
+            version = 'v4';
         }
-        res.redirect(307, versionlessURL);
+
+        // Rewrite the url
+        req.url = versionlessUrl;
+
+        // Add the accept-version header so our internal systems will act as if it was set on the request
+        req.headers['accept-version'] = `${version}.0`;
```

Comment:
I think we should be overriding the `accept-header` value _only if_ it's not present in the request. The clients can set a more specific minor version through the SDKs, so we'd be loosing valuable information. Not sure what the usecase is overriding it here?

Reply:
Yeah fair, I was thinking that if you're requesting a versioned URL then `accept-version` shouldn't be set, but I guess there are change-over cases where that's not true like Zapier.

Updated.